control/controlclient/map.go

```diff
@@ -343,6 +345,40 @@ func undeltaPeers(mapRes *tailcfg.MapResponse, prev []*tailcfg.Node) {
 	mapRes.PeersRemoved = nil
 }
 
+// For extra defense-in-depth, when we're testing expired nodes we check
+// ControlTime against this 'epoch' (set to the approximate time that this code
+// was written) such that if control (or Headscale, etc.) sends a ControlTime
+// that's sufficiently far in the past, we can safely ignore it.
+var removeExpiredPeersEpoch = time.Unix(1673373066, 0)
+
+// removeExpiredPeers updates mapRes.Peers to remove all peers that have
+// expired per the ControlTime field in the MapResponse.
+//
+// This is a defense-in-depth against something going wrong with control such
+// that we start seeing expired peers; we'll drop them here, if so.
+func removeExpiredPeers(mapRes *tailcfg.MapResponse) {
+	if mapRes.ControlTime == nil {
+		log.Printf("netmap: removeExpiredPeers: MapResponse.ControlTime not present; skipping")
+		return
+	}
```

Comment:
this is going to be spammy and also insufficient: most map responses don't have ControlTime set.

only the rare full path adds that.

and we don't really want those extra bytes all the time for every little update. ControlTime is supposed to be an occasional adjustment thing so the client can learn its offset.

Reply:
Storing the difference between control time and local time whenever the field _is_ set may be a close enough approximation, especially if we arrange for it to be sent at least once a day or something. If the difference is small enough (e.g. less than a minute), we could also simply say the local clock is fine and use local time.

Reply:
If we never get a ControlTime (does headscale send it?) then I'd recommend just assuming the local time is good.

Reply:
Okay, I reworked this; we now store a delta on the `mapSession` and use that to calculate the adjusted current time before checking expiry. Thoughts?

Also: @maisem thinks that we shouldn't drop peers here, and should instead mutate their keys to be invalid; if nobody has any objections I can do that next.

Reply:
I view mutating keys to be invalid as mostly a server-side hack for old clients.

You can _also_ do that for defense in depth, but I'd like to set a "expired bool" on the peer in the netmap so things like "tailscale ping" and magicsock and debug can say/log explicit things about peers being expired rather than obscure wireguard key mismatch errors.

Reply:
dropping peers from the netmap would mean that "tailscale status" and the likes would just omit them. That's probably going to result in more support noise. It also has weird implications in terms of subnet routers.

I'd very much prefer if we just somehow blocked traffic to expired peers instead of completely dropping them.

Reply:
Okay, we now replicate what control does and clear `Endpoints` and `DERP`, clear the node's `Key`, and add an `Expired` boolean and set it to `true` in this codepath.